internal/agent/hooks/kcrypt_uki.go

```diff
@@ -96,10 +97,40 @@ func (k KcryptUKI) Run(c config.Config, _ v1.Spec) error {
 		c.Logger.Infof("Done encrypting %s", p)
 	}
 
+	_, _ = utils.SH("sync")
+
 	err = kcrypt.UnlockAll(true)
 	if err != nil {
 		return err
 	}
+	// Close the unlocked partitions after dealing with them, otherwise we leave them open and they can be mounted by anyone
+	defer func() {
+		for _, p := range append([]string{"COS_OEM", "COS_PERSISTENT"}, c.Install.Encrypt...) {
+			c.Logger.Debugf("Closing unencrypted /dev/disk/by-label/%s", p)
+			out, err := utils.SH(fmt.Sprintf("cryptsetup close /dev/disk/by-label/%s", p))
+			// There is a known error with cryptsetup that it can't close the device because of a semaphore
+			// doesnt seem to affect anything as the device is closed as expected so we ignore it if it matches the
+			// output of the error
+			if err != nil && !strings.Contains(out, "incorrect semaphore state") {
+				c.Logger.Errorf("could not close /dev/disk/by-label/%s: %s", p, out)
+			}
+		}
+	}()
+
+	// Here it can take the oem partition a bit of time to appear after unlocking so we need to retry a couple of time with some waiting
+	// retry + backoff
+	for i := 0; i < 10; i++ {
+		c.Logger.Infof("Waiting for unlocked partition to appear")
+		_, _ = utils.SH("sync")
+		part, _ := utils.SH("blkid -L COS_OEM")
+		if part == "" {
+			c.Logger.Infof("Partition not found, waiting %d seconds before retrying", i)
+			time.Sleep(time.Duration(i) * time.Second)
+			continue
+		}
+		c.Logger.Infof("Partition found, continuing")
+		break
+	}
```

Comment:
We are being optimistic here and assume that we will always find the partition after 10 tries, because we don't handle the other case at all. Compared to the previous code, this at least tries more but maybe we could error out if the partition is not found after all the retries.

(I'm late to the party, the PR is closed, but if you think it's important, we can fix it in another PR)

Reply:
I think this cannot fail in any other case, like this is only finding it after creating, if it fails it should fail before during the creation part, but if that succeeded there is no way the device is not showing up. This is just basically waiting for the kernel to refresh the devices and even then it happens on a very small occasion because we are too fast checking it out